workflow/howto/vercel-ai-sdk.mdx

```diff
@@ -0,0 +1,261 @@
+          console.error("Error in fetch implementation:", error);
+          throw error; // Rethrow error for further handling
+        }
+      }
+    },
+  });
+};
+```
+
+### Using OpenAI client to generate text
+
+Now that we've created the OpenAI client, we can use it to generate text or embeddings etc.
+
+For that, we're going to create a new workflow endpoint that uses the payload as prompt to generate text using the OpenAI client.
+
+```typescript {7, 15-19}
+import { generateText } from 'ai';
+import { serve } from "@upstash/workflow/nextjs";
+import { WorkflowAbort } from '@upstash/workflow';
+import { createWorkflowOpenAI } from './model';
+
+export const { POST } = serve(async (context) => {
+  const openai = createWorkflowOpenAI(context);
+
+  // Important: Must get prompt in a step before using generateText 
+  const prompt = await context.run("get prompt", async () => {
+    return context.requestPayload.prompt;
+  });
+
+  try {
+    const result = await generateText({
+      model: openai('gpt-3.5-turbo'),
+      maxTokens: 2048,
+      prompt,
+    });
+
+    await context.run("text", () => {
+      console.log(`TEXT: ${result.text}`);
+      return result.text;
+    });
+    
+  } catch (error) {    
+    if (error instanceof ToolExecutionError && error.cause instanceof WorkflowAbort) {
+      throw error.cause;
+    } else {
+      throw error;
+    }
+  }
+}, {
+  retries: 0
+});
+```
+
+### Advanced Implementation with Tools
+
+Tools allow the AI model to perform specific actions during text generation. You can learn more about tools in the [Vercel AI SDK documentation](https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling).
+
+When using tools with Upstash Workflow, each tool execution must be wrapped in a workflow step.
+
+```typescript {17-23}
+export const { POST } = serve(async (context) => {
+  const openai = createWorkflowOpenAI(context);
+
+  const prompt = await context.run("get prompt", async () => {
+    return context.requestPayload.prompt;
+  });
+
+  try {
+    const result = await generateText({
+      model: openai('gpt-3.5-turbo'),
+      tools: {
+        weather: tool({
+          description: 'Get the weather in a location',
+          parameters: z.object({
+            location: z.string().describe('The location to get the weather for'),
+          }),
+          execute: ({ location }) => context.run("weather tool", () => {
+			// Mock data, replace with actual weather API call
+            return {
+              location,
+              temperature: 72 + Math.floor(Math.random() * 21) - 10,
+            };
+          })
+        }),
+      },
+      maxSteps: 2,
+      prompt,
+    });
+    
+    await context.run("text", () => {
+      return result.text;
+    });
+  } catch (error) {
+    if (error instanceof ToolExecutionError && error.cause instanceof WorkflowAbort) {
+      throw error.cause;
+    } else {
+      throw error;
+    }
+  }
+});
+```
+
+## Important Considerations
```

Comment:
we should say something about ai sdk version somewhere

[it should be at least 4.0.12 because of ToolExecutionError](https://github.com/vercel/ai/releases/tag/ai%404.0.12)